addons/addon-base-raas/packages/base-raas-services/lib/environment/environment-mount-service.js

```diff
@@ -212,15 +235,406 @@ class EnvironmentMountService extends Service {
     ]);
   }
 
+  /**
+   * Function that calculates which users need should and should not have access permissions for the given studyId
+   * Accordingly calls methods that ensure/remove/update permissions for said users
+   *
+   * @param {String} studyId
+   * @param {Object} updateRequest - permission add/remove/update requests coming from SWB UI
+   */
+  async applyWorkspacePermissions(studyId, updateRequest) {
+    const allowedUsers = this._getAllowedUsers(updateRequest);
+    const disAllowedUsers = this._getDisAllowedUsers(updateRequest);
+    const permissionChangeUsers = this._getPermissionChangeUsers(updateRequest);
+
+    if (allowedUsers.length) {
+      await this.addPermissions(allowedUsers, studyId);
+    }
+    if (disAllowedUsers.length) {
+      await this.removePermissions(disAllowedUsers, studyId);
+    }
+    if (permissionChangeUsers.length) {
+      await this.updatePermissions(permissionChangeUsers, studyId, updateRequest);
+    }
+  }
+
+  /**
+   * This method will add/ensure ReadOnly or Read/Write access for workspaces owned by users in the given list
+   * only if these workspaces have this study mounted on it during provision time.
+   * This will not mount studies on user-owned workspaces which did not have them
+   *
+   * @param {Object[]} allowedUsers - Users that newly/continue-to have access to given studyId
+   * @param {String} studyId
+   */
+  async addPermissions(allowedUsers, studyId) {
+    const [environmentScService, iamService] = await this.service(['environmentScService', 'iamService']);
+    await Promise.all(
+      _.map(allowedUsers, async user => {
+        const userOwnedEnvs = await environmentScService.getActiveEnvsForUser(user.uid);
+        const envsWithStudy = _.filter(userOwnedEnvs, env => _.includes(env.studyIds, studyId));
+        await Promise.all(
+          _.map(envsWithStudy, async env => {
+            const {
+              iamClient,
+              studyPathArn,
+              policyDoc,
+              roleName,
+              studyDataPolicyName,
+            } = await this._getIamUpdateParams(env, studyId);
+
+            const statementSidToUse = this._getStatementSidToUse(user.permissionLevel);
+            policyDoc.Statement = this._getStatementsAfterAddition(policyDoc, studyPathArn, statementSidToUse);
+            policyDoc.Statement = this._ensureListAccess(policyDoc, studyPathArn);
+            await iamService.putRolePolicy(roleName, studyDataPolicyName, JSON.stringify(policyDoc), iamClient);
```

Comment:
For all the role policies which are being updated, do we need to hold a lock in order to avoid race conditions and simultaneous updates to the policy. We do something similar when updating S3 bucket policies.

Reply:
I was wondering if that should be P1 or P2. I guess it makes sense for P1 too.